# RNN(LSTM)
<!--<h1 style="text-align:center;">RNN(LSTM)</h1>-->

In [37]:
%env KERAS_BACKEND=tensorflow

env: KERAS_BACKEND=tensorflow


## import keras packages

In [71]:
from keras.models import Sequential
from keras.layers import LSTM, Embedding, Dense, Input, Dropout
from keras.optimizers import Adam

## import dataset IMDB

In [38]:
from keras.datasets import imdb

In [43]:
(x_train0, y_train0), (x_test0, y_test0) = imdb.load_data(num_words=9487)

17465344/17464789 [==============================] - 15s 1us/step


## resize data

In [47]:
from keras.preprocessing import sequence

In [64]:
x_train = sequence.pad_sequences(x_train0, maxlen = 123)
x_test = sequence.pad_sequences(x_test0, maxlen = 123)

## construct RNN

In [90]:
model = Sequential([Embedding(9487, 25),
                    Dropout(0.45), # 看起來很厲害 我也要加
                    LSTM(10),
                    Dropout(0.25), # 看起來很厲害 我也要加
                    Dense(1, activation='sigmoid')])
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_18 (Embedding)     (None, None, 25)          237175    
_________________________________________________________________
dropout_11 (Dropout)         (None, None, 25)          0         
_________________________________________________________________
lstm_13 (LSTM)               (None, 10)                1440      
_________________________________________________________________
dropout_12 (Dropout)         (None, 10)                0         
_________________________________________________________________
dense_29 (Dense)             (None, 1)                 11        
Total params: 238,626
Trainable params: 238,626
Non-trainable params: 0
_________________________________________________________________


In [91]:
model.compile(loss='binary_crossentropy', 
              optimizer=Adam(lr = 0.0009487),
              metrics=['acc'])

In [92]:
model.fit(x_train, y_train0, batch_size=50, epochs=3,
          validation_data = (x_test, y_test0))

Train on 25000 samples, validate on 25000 samples
Epoch 1/3
25000/25000 [==============================] - 56s 2ms/step - loss: 0.5246 - acc: 0.7441 - val_loss: 0.3662 - val_acc: 0.8479
Epoch 2/3
25000/25000 [==============================] - 51s 2ms/step - loss: 0.3109 - acc: 0.8797 - val_loss: 0.3344 - val_acc: 0.8562
Epoch 3/3
25000/25000 [==============================] - 51s 2ms/step - loss: 0.2509 - acc: 0.9079 - val_loss: 0.3368 - val_acc: 0.8564


# summary

1. The trainning took lot of time but still can't do well on validation data(85.64% is my highest score QQ).